ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java

```diff
@@ -118,6 +139,78 @@ ExecuteResult execute(final KsqlPlan plan) {
     return ExecuteResult.of(executePersistentQuery(queryPlan, plan.getStatementText()));
   }
 
+  PullQueryResult executePullQuery(
+      final KsqlExecutionContext ksqlEngine,
+      final ConfiguredStatement<Query> statement,
+      final RoutingFilterFactory routingFilterFactory,
+      final Map<String, Object> requestProperties,
+      final Optional<Boolean> isInternalRequest,
+      final Optional<PullQueryExecutorMetrics> pullQueryMetrics
+  ) {
+
+    if (!statement.getStatement().isPullQuery()) {
+      throw new IllegalArgumentException("Executor can only handle pull queries");
+    }
+    final SessionConfig sessionConfig = statement.getSessionConfig();
+    if (!sessionConfig.getConfig(false)
+        .getBoolean(KsqlConfig.KSQL_PULL_QUERIES_ENABLE_CONFIG)) {
+      throw new KsqlStatementException(
+          "Pull queries are disabled."
+              + PullQueryValidator.PULL_QUERY_SYNTAX_HELP
+              + System.lineSeparator()
+              + "Please set " + KsqlConfig.KSQL_PULL_QUERIES_ENABLE_CONFIG + "=true to enable "
+              + "this feature.",
+          statement.getStatementText());
+    }
+
+    final RoutingOptions routingOptions = new ConfigRoutingOptions(
+        sessionConfig.getConfig(true),
+        requestProperties
+    );
+    final RateLimiter rateLimiter = RateLimiter.create(sessionConfig.getConfig(true).getInt(
+        KsqlConfig.KSQL_QUERY_PULL_MAX_QPS_CONFIG));
+    // If internal listeners are in use, we require the request to come from that listener to
+    // treat it as having been forwarded.
+    final boolean isAlreadyForwarded = routingOptions.skipForwardRequest()
+        // Trust the forward request option if isInternalRequest isn't available.
+        && isInternalRequest.orElse(true);
+
+    // Only check the rate limit at the forwarding host
+    if (!isAlreadyForwarded) {
+      checkRateLimit(rateLimiter);
+    }
+
+
+    try {
+      final QueryAnalyzer queryAnalyzer = new QueryAnalyzer(engineContext.getMetaStore(), "");
+      final ImmutableAnalysis analysis = new RewrittenAnalysis(
+          queryAnalyzer.analyze(statement.getStatement(), Optional.empty()),
+          new ColumnReferenceRewriter()::process
+      );
```

Comment:
Is this not the type of logic that would live in the logical planner? Or at least in `buildAndValidateLogicalPlan`

Reply:
The analyzer is a different step from the logical plan.

Reply:
I thought the plan is to have a single code path for logical plans across all queries. But this new function is only for pull queries, whereas for other queries we are still using `QueryEngine#buildQueryLogicalPlan`.

Reply:
Yes but they are doing the same thing with respect to the logical planner i.e. calling the same methods to build the logical plan. The difference is in the Analyzer where for pull queries there is something different done, that's why I had to create a new method. The analyzer part will be cleaned up in a later PR once we get more clarity about what we need from push queries as well.